src/TM/Engine.java

```diff
@@ -0,0 +1,122 @@
+    String line;
+    int[] pos = new int[1];
+    while ((line = br.readLine()) != null) {
+      pos[0] = 0;
+      Instruction instruction = new Instruction();
+      switch (line.charAt(pos[0])) {
+        case 'b':
+          if (line.charAt(pos[0] + 5) == '(') {
+            instruction.type = InstructionType.BEGIN;
+          } else {
+            instruction.type = InstructionType.BEGINRO;
+          }
+          instruction.transactionIndex = parseIndex(line, pos);
+          break;
+        case 'W':
+          instruction.type = InstructionType.W;
+          instruction.transactionIndex = parseIndex(line, pos);
+          instruction.variableIndex = parseIndex(line, pos);
+          instruction.value = parseIndex(line, pos);
+          break;
+        case 'e':
+          instruction.type = InstructionType.END;
+          instruction.transactionIndex = parseIndex(line, pos);
+          break;
+        case 'd':
+          instruction.type = InstructionType.DUMP;
+          if (line.charAt(pos[0] + 5) == 'x') {
+            instruction.variableIndex = parseIndex(line, pos);
+          } else if (isDigit(line.charAt(pos[0] + 5))) {
+            instruction.value = parseIndex(line, pos);
+          }
+          break;
+        case 'R':
+          instruction.type = InstructionType.R;
+          instruction.transactionIndex = parseIndex(line, pos);
+          instruction.variableIndex = parseIndex(line, pos);
+          break;
+        case 'f':
+          instruction.type = InstructionType.FAIL;
+          instruction.value = parseIndex(line, pos);
+          break;
+        case 'r':
+          instruction.type = InstructionType.RECOVER;
+          instruction.value = parseIndex(line, pos);
+      }
+      instructionList.add(instruction);
+    }
+    if (verbose) {
+      for (Instruction instruction : instructionList) {
+        System.out.println(instruction);
+      }
+    }
+    br.close();
+  }
+
+  private void parseInstructions(boolean verbose) {
+    for (Instruction instruction : instructionList) {
+      switch (instruction.type) {
+        case BEGIN:
+          transactionList.add(new Transaction(instruction.transactionIndex, false));
+          break;
+        case BEGINRO:
+          transactionList.add(new Transaction(instruction.transactionIndex, true));
+          break;
+        case END: case R: case W:
+          transactionList.get(instruction.transactionIndex - 1).instructionList.add(instruction);
+          break;
+      }
+    }
+    if (verbose) {
+      for (Transaction transaction : transactionList) {
+        System.out.println(transaction);
+      }
+    }
+  }
+
+  private void nextDigit(String line, int[] pos) {
+    while (pos[0] < line.length() && !isDigit(line.charAt(pos[0]))) {
+      pos[0]++;
+    }
+  }
+
+  private boolean isDigit(char c) {
+    return c >= '0' && c <= '9';
+  }
+
+  private int parseIndex(String line, int[] pos) {
```

Comment:
Maybe use Integer.parseInt(substring).
We get substring by checking isDigits(charAt(someWhereInTheString)).

https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#parseInt(java.lang.String)